qa/tasks/deepsea.py

```diff
@@ -851,36 +855,76 @@ def _run_orch(self, orch_tuple):
                 cmd_str += ' 2>/dev/null'
         if self.dev_env:
             cmd_str = 'DEV_ENV=true ' + cmd_str
-        try:
-            self.master_remote.run(args=[
-                'sudo', 'bash', '-c', cmd_str,
-                ])
-        except CommandFailedError:
-            self.log.error(anchored(
-                "orchestration {} failed. ".format(orch_spec)
-                + "Here comes journalctl!"
-                ))
-            self.master_remote.run(args=[
-                'sudo',
-                'journalctl',
-                '--all',
-                ])
-            raise
-        self.log.info("orchestration {} completed".format(orch_spec))
+        tries = 0
+        if self.survive_reboots:
+            tries = 5
+        self._exec(cmd_str, orch_spec, tries=tries)
+
+    def _exec(self, cmd_str, orch_spec, tries=0):
+        """
+        Execute cmd_str and catch CommandFailedError and
+        ConnectionLostError and retry to run the command
+        until one of the conditons are fulfilled:
+        1) Execution succeeded
+        2) Attempts are exceeded
+        3) CommandFailedError is raised
+        """
+        with safe_while(sleep=60,
+                        tries=tries,
+                        action="wait for reconnect") as proceed:
+            while proceed():
+                try:
+                    self.master_remote.run(args=[
+                        'sudo', 'bash', '-c', cmd_str,
+                        ])
+                    break
+                except CommandFailedError:
+                    log.error(
+                        "WWWW: orchestration {} failed. ".format(orch_spec)
+                        + "Here comes journalctl!"
+                        )
+                    self.master_remote.run(args=[
+                        'sudo',
+                        'journalctl',
+                        '--all',
+                        ])
+                    raise
+                except ConnectionLostError:
+                    if tries < 1:
+                        raise
+                    log.warning("No connection established yet..")
+
+    def _detect_reboots(self):
+        """
+        Check for all known states/stages/alt-defaults that
+        may cause a reboot
+        If there is a 'allow_reboot' flag, it takes presedence.
+        """
+        orchs_prone_to_reboot = ['ceph.maintenance.upgrade']
+
+        if self.allow_reboot is not None:
+            log.info("Setting allow_reboot explicitly to {}".
+                     format(self.allow_reboot))
+            return self.allow_reboot
+
+        for cnf in self.alternative_defaults:
+            for k, v in cnf.items():
+                if 'reboot' in v and 'no-reboot' not in v:
+                    log.warning("Current orchestration may trigger a reboot")
+                    return True
+
+        if self.state_orch in orchs_prone_to_reboot:
+            log.warning("Current orchestration may trigger a reboot")
```

Comment:
Please change `log.warning` to `self.log.warning` (the former does not set the prefix properly).